### ISSUE TITLE
Input handling issues with VirtualBox

### ISSUE THREAD

**Issue by jacobgkau** [pop-os/cosmic-comp#449]
When using VirtualBox in COSMIC Epoch, I sometimes find my keyboard input still going to VirtualBox and not to other apps after escaping from VirtualBox and switching to another workspace.

This has occurred both with mouse pointer integration enabled and disabled (the latter meaning I have to click inside of the VM/use the right-ctrl key to escape in order to capture/release input). However, when it happens, switching from one to the other seems to work around it for a short time.

**Comment by jacobgkau** [comment]
Probably related, keyboard input sometimes doesn't work when mouse pointer integration is enabled.

**Comment by jacobgkau** [comment]
Also related, I can't use Super key combinations within the VM when the host is running COSMIC Epoch because the host still incorrectly receives the Super keypress (e.g. pressing Super while captured in the VM opens the launcher on the host, and then seems to leave the VM thinking I'm holding the Super key down when I'm not).

**Comment by WaltWeiff** [comment]
Additionally, keyboard input is still captured by VirtualBox after locking the screen -- this makes it impossible to log back in.

**Comment by universal** [comment]
> Additionally, keyboard input is still captured by VirtualBox after locking the screen -- this makes it impossible to log back in.

hitting host-key (right ctrl by default) repeatedly and clicking on the login field helps sometimes...

**Comment by jacobgkau** [comment]
Turning off the "Auto Capture Keyboard" option in Input -> Keyboard -> Keyboard Settings seems to help mitigate this issue. The description of that setting is:

> When checked, the keyboard is automatically captured every time the VM window is activated.

It looks like a proper fix for the problem on the compositor side is very close, but I wanted to mention that workaround.

**Comment by Drakulix** [comment]
Is this still an issue?

**Comment by jacobgkau** [comment]
Yes, this one's still easy to replicate by booting a VM and then switching to a different workspace and trying to type.

**Comment by jacobgkau** [comment]
Yes, this one's still an issue, and easy to replicate by starting a VM (with mouse pointer integration enabled) and then switching to a different workspace and trying to type. (Just reproduced on 0095b6d.)

It does look like Super key handling's improved recently. I can now use Super in the machine without it being sent to the host, although un-capturing my keyboard and then using it on the host still sends it to the guest when it shouldn't.

**Comment by Drakulix** [comment]
> Yes, this one's still an issue, and easy to replicate by starting a VM (with mouse pointer integration enabled) and then switching to a different workspace and trying to type. (Just reproduced on [0095b6d](https://github.com/pop-os/cosmic-comp/commit/0095b6d505fe45e7e09f980cbf48fff1800a9d79).)

So this seems to be multiple issues:

- If the keyboard is grabbed and the workspace is switched via the mouse (e.g. clicking on the workspace applet), focus is not removed from the VM.
- If the keyboard isn't grabbed and the workspace is switched via shortcuts or the mouse, focus doesn't switch correctly, if the workspace is empty. Causing the VM to still receive keyboard input until something else is explicitly focused.
- If no keyboard grab is active and the new workspace contains a window, everything seems to behave as expected.

I'll look into each of them.

> It does look like Super key handling's improved recently. I can now use Super in the machine without it being sent to the host, although un-capturing my keyboard and then using it on the host still sends it to the guest when it shouldn't.

I am not sure, this is a bug. Shortcuts like switching the workspace seem to work correctly and don't get forwarded to the VM. It is only modifier-only shortcuts, that seem to trigger on both the host and the VM.

Which is understandable, because we need to update the client's keymap to contain any pressed modifiers, even if we aren't sure, if the current press might lead to a shortcut. So if the VM window is focused, it'll receive the Super-key events and thus trigger the shortcut inside the VM as well.

Given we filter out any key, that ended up triggering a shortcut, any key-binding containing "Super" will trigger the Super-shortcut inside the VM, if the window was focused previously. Which is indeed quite annoying, but I don't see a good way around this.

(A way to potentially avoid this while still being in spec would be to send a keyboard-leave event before sending the Super-key "Release" event. But that would be racy, because we don't know if the shortcut will trigger a new client, which will take over focus. Of course the Super-shortcut does by default trigger the launcher, but until that window has actually appeared, we already processed the release-event as normal.)

Do you see any other key-combinations misbehaving here?

**Comment by Drakulix** [comment]
https://github.com/pop-os/cosmic-comp/pull/1585 should address the issues around workspace switching not releasing the grab as expected.
